### PR TITLE
`not_unsafe_ptr_arg_deref` update documentation

### LINKED ISSUE #7714: Improve explanation of not_unsafe_ptr_arg_deref
I pointed out in https://github.com/rust-lang/rust-clippy/issues/7666#issuecomment-919006725 that the use of `allow(clippy::not_unsafe_ptr_arg_deref)` on crates.io is shockingly high.

Here is the list: https://dtolnay.github.io/noisy-clippy/not_unsafe_ptr_arg_deref.html

I spot checked some of these and pretty much every case is unsound.

- https://github.com/dtn7/bp7-rs/issues/2
- https://github.com/UnitedTraders/aeron-rs/issues/14
- https://github.com/UnitedTraders/aeron-rs/issues/16
- https://github.com/UnitedTraders/aeron-rs/issues/17
- https://github.com/drogue-iot/drogue-ffi-compat/issues/1
- https://github.com/jni-rs/jni-rs/issues/331
- https://github.com/wiomoc/mosquitto-jwt-auth/issues/11
- https://github.com/pact-foundation/pact-reference/issues/148
- https://github.com/EmbarkStudios/spirv-tools-rs/issues/27
- https://github.com/JoNil/tm-rs/issues/1
- https://github.com/zombodb/pgx/issues/196
- https://github.com/zombodb/pgx/issues/215

If people are not understanding what this lint is attempting to tell them, it seems like the lint isn't doing its job. We should look into whether the lint's documentation or the lint's diagnostic output can be improved in a way that better comes across to people. It may be worth looping in some of the maintainers of some of the crates above about what might have been helpful for them to understand the lint.

### PR DESCRIPTION
**Pull request by tgross35**
changelog: [`not_unsafe_ptr_arg_deref`]: strengthened documentation wording, fixes #7714

### CODE REVIEW

File: clippy_lints/src/functions/mod.rs [hunk 1]
```diff
@@ -63,30 +63,54 @@ declare_clippy_lint! {
...
+    /// // foo(invalid_ptr);
     /// ```
     ///
     /// Use instead:
     /// ```rust,ignore
     /// pub unsafe fn foo(x: *const u8) {
     ///     println!("{}", unsafe { *x });
     /// }
+    ///
+    /// // this call would cause a compiler error
```
Comment (llogiq) [inline clippy_lints/src/functions/mod.rs:109]: I'd add "because of the missing `unsafe`" to be absolutely clear.

**Review by llogiq — COMMENTED** [review]
I like it. I only have one small suggestion, otherwise it's good to merge.

File: clippy_lints/src/functions/mod.rs (hunk 1 above)
Comment (tgross35) [reply clippy_lints/src/functions/mod.rs]: Easy enough, I updated this

**Comment by llogiq** [comment]
Thank you!

@bors r+

**Comment by tgross35** [comment]
Awesome, thanks!

**Pull request merged**